File: ios/Runner/Native/ConnectivityPigeonStreamHandler.swift
import Network

final class ConnectivityPigeonStreamHandler: ConnectivityEventsStreamHandler {
    private var monitor: NWPathMonitor?
    private var eventSink: PigeonEventSink<ApiConnectivityStatus>?

    override func onListen(
        withArguments arguments: Any?,
        sink: PigeonEventSink<ApiConnectivityStatus>
    ) {
        stopMonitor()
        eventSink = sink
        let monitor = NWPathMonitor()
        self.monitor = monitor
        monitor.pathUpdateHandler = { [weak self] path in
            DispatchQueue.main.async {
                self?.eventSink?.success(AppConnectivityStatus.from(path: path).pigeonStatus)
            }
        }
        monitor.start(queue: DispatchQueue(label: "pdp.connectivity.pigeon"))
    }

    override func onCancel(withArguments arguments: Any?) {
        stopMonitor()
        eventSink = nil
    }

    private func stopMonitor() {
        monitor?.cancel()
        monitor = nil
    }
}

File: android/app/src/main/kotlin/com/example/pdp_todo_app/native/ConnectivityPigeonStreamHandler.kt
package com.example.pdp_todo_app.native

import android.content.Context
import android.net.ConnectivityManager
import android.net.Network
import android.net.NetworkCapabilities
import android.net.NetworkRequest
import android.os.Handler
import android.os.Looper
import com.example.pdp_todo_app.pigeon.ApiConnectivityStatus
import com.example.pdp_todo_app.pigeon.ConnectivityEventsStreamHandler
import com.example.pdp_todo_app.pigeon.PigeonEventSink

class ConnectivityPigeonStreamHandler(
    context: Context
) : ConnectivityEventsStreamHandler() {

    private val connectivityManager =
        context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
    private val mainHandler = Handler(Looper.getMainLooper())

    private var eventSink: PigeonEventSink<ApiConnectivityStatus>? = null
    private var callback: ConnectivityManager.NetworkCallback? = null

    override fun onListen(p0: Any?, sink: PigeonEventSink<ApiConnectivityStatus>) {
        unregisterCallback()
        eventSink = sink
        emitStatus()

        val networkCallback = object : ConnectivityManager.NetworkCallback() {
            override fun onAvailable(network: Network) = emitStatus()

            override fun onLost(network: Network) = emitStatus()

            override fun onCapabilitiesChanged(
                network: Network,
                networkCapabilities: NetworkCapabilities
            ) = emitStatus()
        }
        callback = networkCallback
        connectivityManager.registerNetworkCallback(
            NetworkRequest.Builder().build(),
            networkCallback
        )
    }

    override fun onCancel(p0: Any?) {
        unregisterCallback()
        eventSink = null
    }

    fun dispose() {
        onCancel(null)
    }

    private fun unregisterCallback() {
        callback?.let(connectivityManager::unregisterNetworkCallback)
        callback = null
    }

    private fun emitStatus() {
        val status = connectivityManager.appConnectivityStatus().toPigeon()
        mainHandler.post {
            eventSink?.success(status)
        }
    }
}

private fun AppConnectivityStatus.toPigeon(): ApiConnectivityStatus {
    return when (this) {
        AppConnectivityStatus.WIFI -> ApiConnectivityStatus.WIFI
        AppConnectivityStatus.MOBILE -> ApiConnectivityStatus.MOBILE
        AppConnectivityStatus.NONE -> ApiConnectivityStatus.NONE
    }
}
